src/window.c

```diff
@@ -4416,6 +4418,13 @@ close_others(
     for (wp = firstwin; win_valid(wp); wp = nextwp)
     {
 	nextwp = wp->w_next;
+
+	// autocommands messed this one up
+	if (old_curwin != curwin && win_valid(old_curwin))
+	    curwin = old_curwin;
+	if (old_curbuf != curbuf && buf_valid(old_curbuf))
+	    curbuf = old_curbuf;
```

Comment:
This still doesn't look right. If an autocommand (perhaps `BufUnload`) closed `old_curwin` while `old_curbuf` is still loaded in another window, or changed `old_curwin->w_buffer`, this can again lead to `curbuf != curwin->w_buffer`.